src/System.Private.Interpreter/src/Internal/Runtime/Interpreter/StackItem.cs

```diff
@@ -3,68 +3,137 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
+using System.Runtime.InteropServices;
 using Internal.IL;
 using Internal.TypeSystem;
 
 namespace Internal.Runtime.Interpreter
 {
-    internal abstract class StackItem
+    [StructLayout(LayoutKind.Explicit)]
+    internal unsafe struct StackItem
     {
-        public StackValueKind Kind { get; protected set; }
-        public WellKnownType Type { get; protected set; }
-    }
+        [FieldOffset(0)]
+        private StackValueKind _kind;
 
-    internal class StackItem<T> : StackItem
-    {
-        public T Value { get; }
+        [FieldOffset(8)]
+        private WellKnownType _type;
+
+        [FieldOffset(16)]
+        private int _int32;
+
+        [FieldOffset(16)]
+        private long _int64;
+
+        [FieldOffset(16)]
+        private IntPtr _nativeInt;
+
+        [FieldOffset(16)]
+        private float _float;
+
+        [FieldOffset(16)]
+        private double _double;
 
-        public StackItem(T value, StackValueKind kind, WellKnownType type)
+        [FieldOffset(16)]
+        private void* _ptr;
+
+        [FieldOffset(24)]
+        private ValueType _valueType;
+
+        [FieldOffset(24)]
+        private object _objref;
+
+        public StackValueKind Kind => _kind;
+        public WellKnownType Type => _type;
+
+        public static implicit operator StackItem(int int32)
```

Comment:
I would stay away from operators in this type.

I like the `From*` methods you have created. I would use them for everything.